libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/arrow/api/MapboxRouteArrowView.kt

```diff
@@ -23,43 +30,128 @@ import com.mapbox.navigation.ui.maps.route.arrow.model.RouteArrowState
-     * it contains.
+     * Renders a Expected<UpdateManeuverArrowValue, RouteArrowUpdateError> applying view side
+     * effects based on the data it contains.
      *
      * @param style a valid map style object
      * @param state a state containing data for applying the view side effects.
      */
-    fun render(style: Style, state: RouteArrowState.UpdateManeuverArrowState) {
+    fun renderManeuverUpdate(
+        style: Style,
+        state: Expected<UpdateManeuverArrowValue, RouteArrowUpdateError>
+    ) {
         initializeLayers(style, options)
 
-        state.getVisibilityChanges().forEach {
-            updateLayerVisibility(style, it.first, it.second)
+        when (state) {
+            is Expected.Failure -> {
+            }
+            is Expected.Success -> {
+                state.value.layerVisibilityModifications.forEach {
+                    updateLayerVisibility(style, it.first, it.second)
+                }
+                state.value.arrowHeadFeature?.apply {
+                    updateSource(style, RouteConstants.ARROW_HEAD_SOURCE_ID, this)
+                }
+                state.value.arrowShaftFeature?.apply {
+                    updateSource(style, RouteConstants.ARROW_SHAFT_SOURCE_ID, this)
+                }
+            }
         }
+    }
 
-        state.getArrowHeadFeature()?.apply {
-            updateSource(style, RouteConstants.ARROW_HEAD_SOURCE_ID, this)
-        }
+    /**
+     * Renders a [ArrowAddedValue]
+     *
+     * @param style a valid map style object
+     * @param arrowAdded a state containing data for applying the view side effects.
+     */
+    fun render(style: Style, arrowAdded: ArrowAddedValue) {
+        initializeLayers(style, options)
 
-        state.getArrowShaftFeature()?.apply {
-            updateSource(style, RouteConstants.ARROW_SHAFT_SOURCE_ID, this)
+        updateSource(
+            style,
+            RouteConstants.ARROW_SHAFT_SOURCE_ID,
+            arrowAdded.arrowShaftFeatureCollection
+        )
+        updateSource(
+            style,
+            RouteConstants.ARROW_HEAD_SOURCE_ID,
+            arrowAdded.arrowHeadFeatureCollection
+        )
+    }
+
+    /**
+     * Renders a [ArrowAddedValue]
+     *
+     * @param style a valid map style object
+     * @param expectedValue a state containing data for applying the view side effects.
+     */
+    fun render(style: Style, expectedValue: Expected<ArrowAddedValue, RouteArrowUpdateError>) {
+        when (expectedValue) {
+            is Expected.Failure -> { }
+            is Expected.Success -> {
+                render(style, expectedValue.value)
+            }
         }
     }
 
+    /**
+     * Renders a [RemoveArrowValue]
+     *
+     * @param style a valid map style object
+     * @param state a state containing data for applying the view side effects.
+     */
+    fun render(style: Style, state: RemoveArrowValue) {
```

Comment:
Should we have 3 different `render` versions for each different state (`ArrowAddedValue`, `RemoveArrowValue` and `ClearArrowsValue`)? If so, we should at least extract that piece of code into a `private` `fun`ction and reuse it across them.

Reply:
Each value class represents the return type for each API call. If I combine them that's going to cause SEMVER issues in the future if they diverge.